### PR TITLE
Suggest matching commands if the user makes a typo

### LINKED ISSUE #356: Show up suitable messages if the user makes a typo.
As per now `vsce <unknown_cmd>` only manages to show up the output of `vsce --help`.
It would be nice to have a suitable message such as `unknown command <cmd>` along with a suggestion for matching commands.

# Proposed idea

## Warning for unknown commands.

`vsce random`

```md
Usage: vsce [options] [command]

Options:
  -V, --version                        output the version number
  -h, --help                           output usage information

Commands:
  ls [options]                         Lists all the files that will be published
  package [options]                    Packages an extension
  publish [options] [<version>]        Publishes an extension
  unpublish [options] [<extensionid>]  Unpublishes an extension. Example extension id: microsoft.csharp.
  ls-publishers                        List all known publishers
  create-publisher <publisher>         Creates a new publisher
  delete-publisher <publisher>         Deletes a publisher
  login <publisher>                    Add a publisher to the known publishers list
  logout <publisher>                   Remove a publisher from the known publishers list
  show [options] <extensionid>         Show extension metadata
  search [options] <text>              search extension gallery
  *

Unknown command random
```

## Suggest matching commands if the user makes a typo.

`vsce packag`

```md
Usage: vsce [options] [command]

Options:
  -V, --version                        output the version number
  -h, --help                           output usage information

Commands:
  ls [options]                         Lists all the files that will be published
  package [options]                    Packages an extension
  publish [options] [<version>]        Publishes an extension
  unpublish [options] [<extensionid>]  Unpublishes an extension. Example extension id: microsoft.csharp.
  ls-publishers                        List all known publishers
  create-publisher <publisher>         Creates a new publisher
  delete-publisher <publisher>         Deletes a publisher
  login <publisher>                    Add a publisher to the known publishers list
  logout <publisher>                   Remove a publisher from the known publishers list
  show [options] <extensionid>         Show extension metadata
  search [options] <text>              search extension gallery
  *

Unknown command packag

Did you mean package?
```

### PR DESCRIPTION
**Pull request by jamesgeorge007**
Fixes #356 

Suggests matching commands if the user makes a typo:-

Screenshot from 2019-06-03 22-32-20

### CODE REVIEW

**Comment by jamesgeorge007** [comment]
@joaomoreno Thoughts :thinking:

File: src/main.ts [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import * as program from 'commander';
+const didYouMean = require('didyoumean');
```
Comment (joaomoreno) [inline src/main.ts]: Please use the `import from` syntax, and make sure there are typings for this.

File: package.json [hunk 2]
```diff
@@ -39,6 +39,7 @@
     "cheerio": "^1.0.0-rc.1",
     "commander": "^2.8.1",
     "denodeify": "^1.2.1",
+    "didyoumean": "^1.2.1",
```
Comment (joaomoreno) [inline package.json]: You should use `yarn add didyoumean`, to update `yarn.lock` as well.

File: src/main.ts [hunk 3]
```diff
@@ -127,7 +140,11 @@ module.exports = function (argv: string[]): void {
 
 	program
 		.command('*')
-		.action(() => program.help());
+		.action((cmd: string) => {
+			program.outputHelp();
+			log.error(`Unknown command ${cmd}`);
+			suggestCommands(cmd);
```
Comment (joaomoreno) [inline src/main.ts]: Let's inline this function here.

File: src/main.ts [hunk 4]
```diff
@@ -51,6 +53,17 @@ function main(task: Promise<any>): void {
 		});
 }
 
+function suggestCommands(cmd: string): void {
+	const availableCommands = program.commands.map(cmd => {
+		return cmd._name;
```
Comment (joaomoreno) [inline src/main.ts]: Indentation is broken. Make sure you run `Format Document` before saving.

File: src/main.ts [hunk 5]
```diff
@@ -51,6 +53,17 @@ function main(task: Promise<any>): void {
 		});
 }
 
+function suggestCommands(cmd: string): void {
+	const availableCommands = program.commands.map(cmd => {
```
Comment (joaomoreno) [inline src/main.ts]:
```suggestion
	const availableCommands = program.commands.map(c => c._name);
```

File: src/main.ts (hunk 1 above)
Comment (joaomoreno) [reply src/main.ts]: If there are no official typings, feel free to create a `typings` folder and add `didyoumean.d.ts` to it. Better yet, submit `@types/didyoumean` to NPM and use that!

File: src/main.ts [hunk 6]
```diff
@@ -51,6 +53,17 @@ function main(task: Promise<any>): void {
...
+		return cmd._name;
+	  });
+	
+	  const suggestion: string | string[] = didYouMean(cmd, availableCommands);
+	  if (suggestion) {
+		log.warn(`Did you mean ${suggestion}?`);
```
Comment (joaomoreno) [inline src/main.ts]: Wrap `${suggestion}` in single quotes.

File: src/main.ts (hunk 3 above)
Comment (jamesgeorge007) [reply src/main.ts]: Well, could you elaborate :thinking:

Comment (joaomoreno) [reply src/main.ts]: I think we don't need a function for this, let's just take the body of `suggestCommands` and inline that here.

File: src/main.ts (hunk 1 above)
Comment (jamesgeorge007) [reply src/main.ts]: I think we're good to go after this one gets merged: [#27](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35950)

Comment (joaomoreno) [reply src/main.ts]: :clap: !

File: src/main.ts (hunk 3 above)
Comment (jamesgeorge007) [reply src/main.ts]: :+1:

**Comment by jamesgeorge007** [comment]
@joaomoreno Hope, we're good to go :+1:

**Comment by jamesgeorge007** [comment]
@joaomoreno Hope, we're good to go :+1:

**Comment by joaomoreno** [comment]
Awesome thanks! :beers: 

`yarn.lock` changes were missing again and I've refactored everything just a bit.

**Comment by joaomoreno** [comment]
Released 1.63.0!

**Pull request merged**